File: examples/typescript/part-3/10-secrets-management.ts
// Illustrative examples for notes/part-3-processing-security/10-secrets-management.md
// Not for production use. Licensed under MIT (see LICENSE-CODE).

interface SecretScope {
  toolName: string;
  action: string;
  userId: string;
  resource: string;
}

interface SecretProvider {
  getScopedToken(scope: SecretScope): Promise<string>;
}

interface HTTPClient {
  post(url: string, bearerToken: string, body: Uint8Array): Promise<Uint8Array>;
}

interface SendRequestArgs {
  endpoint: string;
  payload: string;
  // Важно: здесь нет Token/APIKey.
  // LLM не должна формировать секрет в аргументах.
}

class ExternalAPIClient {
  constructor(
    private secrets: SecretProvider,
    private http: HTTPClient,
  ) {}

  async sendRequest(userId: string, args: SendRequestArgs): Promise<Uint8Array> {
    if (!args.endpoint) {
      throw new Error("endpoint is required");
    }

    const token = await this.secrets.getScopedToken({
      toolName: "external_api.send_request",
      action: "send",
      userId,
      resource: args.endpoint,
    });

    return this.http.post(
      args.endpoint,
      token,
      new TextEncoder().encode(args.payload),
    );
  }
}

// --- Redacted logger ---

class Redactor {
  private patterns: RegExp[] = [
    /(api[_-]?key|token|password|secret)\s*[:=]\s*[^\s,]+/gi,
    /-----BEGIN [A-Z ]+PRIVATE KEY-----[\s\S]+?-----END [A-Z ]+PRIVATE KEY-----/g,
    /bearer\s+[a-z0-9._\-]+/gi,
  ];

  redact(text: string): string {
    let out = text;
    for (const pattern of this.patterns) {
      out = out.replace(pattern, "[REDACTED_SECRET]");
    }
    return out;
  }
}

class SafeLogger {
  constructor(private redactor: Redactor) {}

  log(event: string, fields: Record<string, string>): void {
    const safe: Record<string, string> = {};
    for (const [key, value] of Object.entries(fields)) {
      if (looksSensitiveKey(key)) {
        safe[key] = "[REDACTED_SECRET]";
      } else {
        safe[key] = this.redactor.redact(value);
      }
    }

    // Здесь запись в реальный logger/tracing backend.
    void event;
    void safe;
  }
}

function looksSensitiveKey(key: string): boolean {
  const lower = key.toLowerCase();
  const markers = ["token", "password", "secret", "api_key", "authorization"];
  return markers.some((m) => lower.includes(m));
}

export { ExternalAPIClient, Redactor, SafeLogger, looksSensitiveKey };

export type { SecretScope, SecretProvider, HTTPClient, SendRequestArgs };

File: examples/python/part-3/10-secrets-management.py
# Illustrative examples for notes/part-3-processing-security/10-secrets-management.md
# Not for production use. Licensed under MIT (see LICENSE-CODE).

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Protocol


@dataclass
class SecretScope:
    tool_name: str
    action: str
    user_id: str
    resource: str


class SecretProvider(Protocol):
    def get_scoped_token(self, scope: SecretScope) -> str: ...


class HTTPClient(Protocol):
    def post(self, url: str, bearer_token: str, body: bytes) -> bytes: ...


@dataclass
class SendRequestArgs:
    endpoint: str
    payload: str
    # Важно: здесь нет Token/APIKey.
    # LLM не должна формировать секрет в аргументах.


class ExternalAPIClient:
    def __init__(self, secrets: SecretProvider, http: HTTPClient) -> None:
        self.secrets = secrets
        self.http = http

    def send_request(self, user_id: str, args: SendRequestArgs) -> bytes:
        if not args.endpoint:
            raise ValueError("endpoint is required")

        token = self.secrets.get_scoped_token(
            SecretScope(
                tool_name="external_api.send_request",
                action="send",
                user_id=user_id,
                resource=args.endpoint,
            )
        )

        return self.http.post(args.endpoint, token, args.payload.encode("utf-8"))


# --- Redacted logger ---


class Redactor:
    def __init__(self) -> None:
        self.patterns = [
            re.compile(
                r"(?i)(api[_-]?key|token|password|secret)\s*[:=]\s*[^\s,]+",
            ),
            re.compile(
                r"-----BEGIN [A-Z ]+PRIVATE KEY-----[\s\S]+?-----END [A-Z ]+PRIVATE KEY-----",
            ),
            re.compile(r"(?i)bearer\s+[a-z0-9._\-]+"),
        ]

    def redact(self, text: str) -> str:
        out = text
        for pattern in self.patterns:
            out = pattern.sub("[REDACTED_SECRET]", out)
        return out


class SafeLogger:
    def __init__(self, redactor: Redactor) -> None:
        self.redactor = redactor

    def log(self, event: str, fields: dict[str, str]) -> None:
        safe: dict[str, str] = {}
        for key, value in fields.items():
            if _looks_sensitive_key(key):
                safe[key] = "[REDACTED_SECRET]"
            else:
                safe[key] = self.redactor.redact(value)

        # Здесь запись в реальный logger/tracing backend.
        _ = event
        _ = safe


def _looks_sensitive_key(key: str) -> bool:
    lower = key.lower()
    markers = ["token", "password", "secret", "api_key", "authorization"]
    return any(m in lower for m in markers)
